### PR TITLE
Refactored instance metadata and project metadata to share functionality

### PR DESCRIPTION
**Pull request by lwander**

### CODE REVIEW

File: builtin/providers/google/metadata.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	// Load metadata on server into map
+	for _, kv := range serverMD.Items {
+		// If the server state has a key that we had in our old
+		// state, but not in our new state, we should delete it
+		_, okOld := oldMDMap[kv.Key]
+		_, okNew := newMDMap[kv.Key]
+		if okOld && !okNew {
+			continue
+		} else {
+			if kv.Value != nil {
```
Comment (sparkprime) [inline builtin/providers/google/metadata.go]: Can this still be nil?

Comment (lwander) [inline builtin/providers/google/metadata.go]: The [API](https://godoc.org/google.golang.org/api/compute/v1#MetadataItems) doesn't say anything explicitly. Do you think I should remove the check?

File: builtin/providers/google/resource_compute_instance.go [hunk 2]
```diff
@@ -673,23 +681,39 @@ func resourceComputeInstanceUpdate(d *schema.ResourceData, meta interface{}) err
...
+			if err != nil {
+				return fmt.Errorf("Error updating metadata: %s", err)
+			}
+			op, err := config.clientCompute.Instances.SetMetadata(
+				config.Project, zone, d.Id(), md).Do()
+			if err != nil {
+				return fmt.Errorf("Error updating metadata: %s", err)
+			}
+
+			// 1 5 2
```
Comment (sparkprime) [inline builtin/providers/google/resource_compute_instance.go]: What does this comment mean?

File: builtin/providers/google/metadata.go (hunk 1 above)
Comment (sparkprime) [inline builtin/providers/google/metadata.go]: We know from experience that if you actually pass null into the API then it gets upset.  So I think it's reasonable to assume we won't ever get null out again.

File: builtin/providers/google/resource_compute_instance.go (hunk 2 above)
Comment (lwander) [inline builtin/providers/google/resource_compute_instance.go]: I'm not sure, It was in the code before I made changes... It probably doesn't make sense for me to keep it though.

File: builtin/providers/google/metadata.go (hunk 1 above)
Comment (lwander) [inline builtin/providers/google/metadata.go]: Sounds good

**Comment by sparkprime** [comment]
Otherwise LGTM, but it needs merging and squashing

**Comment by phinze** [comment]
Hi @lwander and @sparkprime - I'm getting a bunch of metadata-related errors on master's acceptance tests. Spelunking through the diffs led me to this PR.

Example failure: https://gist.github.com/phinze/3c837b2a71dc2b7cdbd1

Repro command to run:

```
TF_ACC=1 go test -v ./builtin/providers/google -run TestAccComputeInstance_basic$
```

I'll keep digging in but I wanted to give you a heads up in case you see the problem right away.

**Comment by lwander** [comment]
I think I see where the problem is, I'll send you a PR soon.

**Pull request merged**